edweiss-app/components/core/calendar.tsx

```diff
@@ -57,102 +60,190 @@ export const Calendar = ({ courses, type }: { courses: { id: string; data: Cours
         fetchUser();
     }, [userId]);
 
+    // Update current minutes every minute
     useEffect(() => {
-        const interval = setInterval(() => { setCurrentMinutes(getCurrentTimeInMinutes()); }, 60000);
+        const interval = setInterval(() => setCurrentMinutes(getCurrentTimeInMinutes()), 60000);
         return () => clearInterval(interval);
     }, []);
 
+    // Scroll to current time block
     useEffect(() => {
         scrollViewRef.current?.scrollTo({ y: (currentMinutes / 60 - 1) * HOUR_BLOCK_HEIGHT, animated: true });
-    }, []);
+    }, [currentMinutes]);
+
+    // Memoize the week dates
+    const weekDates = useMemo(() => getWeekDates(date), [date]);
+
+    const renderDay = (dayDate: Date) => {
+        const dayIndex = dayDate.getDay();
+
+        // Filter assignments and todos for the given day
+        const dailyAssignments = assignments.filter(assignment => {
+            const assignmentDate = Time.toDate(assignment.data.dueDate);
+            return (
+                assignmentDate.getDate() === dayDate.getDate() &&
+                assignmentDate.getMonth() === dayDate.getMonth() &&
+                assignmentDate.getFullYear() === dayDate.getFullYear()
+            );
+        });
+
+        const dailyTodos = todos?.filter(todo => {
+            const todoDate = todo.data?.dueDate ? Time.toDate(todo.data.dueDate) : null;
+            return (
+                todoDate?.getDate() === dayDate.getDate() &&
+                todoDate?.getMonth() === dayDate.getMonth() &&
+                todoDate?.getFullYear() === dayDate.getFullYear()
+            );
+        });
```

Comment:
Here are small changes to improve the modularity and separation of concerns of your code.

I feel like that the `renderDay` function handles several responsibilities:
- Filtering assignments and todos.
- Filtering course periods and related courses.
- Rendering hour blocks.
This leads to bloated and less readable code. Extracting filtering logic into separate helper functions improves readability and testability.
Moreover using a `usememo` may enhance the performance. Here's what I suggest:

```tsx
const getDailyAssignments = (dayDate: Date) =>
    assignments.filter(assignment => {
        const assignmentDate = Time.toDate(assignment.data.dueDate);
        return (
            assignmentDate.getDate() === dayDate.getDate() &&
            assignmentDate.getMonth() === dayDate.getMonth() &&
            assignmentDate.getFullYear() === dayDate.getFullYear()
        );
    });

const getFilteredPeriods = (courses: { id: string; data: Course }[], dayIndex: number, hour: number) =>
    courses.flatMap(course =>
        course.data.periods.filter(
            period =>
                period.dayIndex === dayIndex &&
                period.start >= hour * 60 &&
                period.start < (hour + 1) * 60
        )
    );

// Inside renderDay
const dailyAssignments = useMemo(() => getDailyAssignments(dayDate), [dayDate, assignments]);
const filteredPeriods = useMemo(() => getFilteredPeriods(courses, dayIndex, hour), [courses, dayIndex, hour]);
```
What do you think about this ?